File: 4_Behavioral_Design_Patterns/4.11_Visitor/Python_Snippet/Visitor.py
# Visitor Interface
from abc import ABC, abstractmethod

# Element Interface
class Animal(ABC):
    @abstractmethod
    def accept(self, visitor):
        pass

# Concrete Element: Lion
class Lion(Animal):
    def accept(self, visitor):
        visitor.visit_lion(self)

# Concrete Element: Elephant
class Elephant(Animal):
    def accept(self, visitor):
        visitor.visit_elephant(self)

# Concrete Element: Giraffe
class Giraffe(Animal):
    def accept(self, visitor):
        visitor.visit_giraffe(self)

class AnimalVisitor(ABC):
    @abstractmethod
    def visit_lion(self, lion: Lion):
        pass

    @abstractmethod
    def visit_elephant(self, elephant: Elephant):
        pass

    @abstractmethod
    def visit_giraffe(self, giraffe: Giraffe):
        pass

# Concrete Visitor: FeedVisitor
class FeedVisitor(AnimalVisitor):
    def visit_lion(self, lion):
        print("Feeding lion")

    def visit_elephant(self, elephant):
        print("Feeding elephant")

    def visit_giraffe(self, giraffe):
        print("Feeding giraffe")

# Concrete Visitor: SleepVisitor
class SleepVisitor(AnimalVisitor):
    def visit_lion(self, lion):
        print("Putting lion to sleep")

    def visit_elephant(self, elephant):
        print("Putting elephant to sleep")

    def visit_giraffe(self, giraffe):
        print("Putting giraffe to sleep")

# Usage
if __name__ == "__main__":
    lion = Lion()
    elephant = Elephant()
    giraffe = Giraffe()

    feed_visitor = FeedVisitor()
    sleep_visitor = SleepVisitor()

    lion.accept(feed_visitor)  # Output: Feeding lion
    elephant.accept(feed_visitor)  # Output: Feeding elephant
    giraffe.accept(feed_visitor)  # Output: Feeding giraffe

    lion.accept(sleep_visitor)  # Output: Putting lion to sleep
    elephant.accept(sleep_visitor)  # Output: Putting elephant to sleep
    giraffe.accept(sleep_visitor)  # Output: Putting giraffe to sleep

File: 4_Behavioral_Design_Patterns/4.11_Visitor/Java_Snippet/Visitor.java
// Visitor Interface
interface AnimalVisitor {
    void visit(Lion lion);

    void visit(Elephant elephant);

    void visit(Giraffe giraffe);
}

// Element Interface
interface Animal {
    void accept(AnimalVisitor visitor);
}

// Concrete Element: Lion
class Lion implements Animal {
    @Override
    public void accept(AnimalVisitor visitor) {
        visitor.visit(this);
    }
}

// Concrete Element: Elephant
class Elephant implements Animal {
    @Override
    public void accept(AnimalVisitor visitor) {
        visitor.visit(this);
    }
}

// Concrete Element: Giraffe
class Giraffe implements Animal {
    @Override
    public void accept(AnimalVisitor visitor) {
        visitor.visit(this);
    }
}

// Concrete Visitor: FeedVisitor
class FeedVisitor implements AnimalVisitor {
    @Override
    public void visit(Lion lion) {
        System.out.println("Feeding lion");
    }

    @Override
    public void visit(Elephant elephant) {
        System.out.println("Feeding elephant");
    }

    @Override
    public void visit(Giraffe giraffe) {
        System.out.println("Feeding giraffe");
    }
}

// Concrete Visitor: SleepVisitor
class SleepVisitor implements AnimalVisitor {
    @Override
    public void visit(Lion lion) {
        System.out.println("Putting lion to sleep");
    }

    @Override
    public void visit(Elephant elephant) {
        System.out.println("Putting elephant to sleep");
    }

    @Override
    public void visit(Giraffe giraffe) {
        System.out.println("Putting giraffe to sleep");
    }
}

// Usage
public class Visitor {
    public static void main(String[] args) {
        Animal lion = new Lion();
        Animal elephant = new Elephant();
        Animal giraffe = new Giraffe();

        AnimalVisitor feedVisitor = new FeedVisitor();
        AnimalVisitor sleepVisitor = new SleepVisitor();

        lion.accept(feedVisitor); // Output: Feeding lion
        elephant.accept(feedVisitor); // Output: Feeding elephant
        giraffe.accept(feedVisitor); // Output: Feeding giraffe

        lion.accept(sleepVisitor); // Output: Putting lion to sleep
        elephant.accept(sleepVisitor); // Output: Putting elephant to sleep
        giraffe.accept(sleepVisitor); // Output: Putting giraffe to sleep
    }
}
